documentation/jetty-documentation/src/main/asciidoc/programming-guide/arch-threads.adoc

```diff
@@ -0,0 +1,200 @@
+
+[[pg-arch-threads-execution-strategy]]
+==== Execution Strategies
+
+The Jetty threading architecture can be modeled with a producer/consumer pattern, where produced tasks needs to be consumed efficiently.
+
+For example, Jetty produces (among others) these tasks:
+
+* A task that wraps a NIO selection event, see the xref:pg-arch-io[Jetty I/O architecture].
+* A task that wraps the invocation of application code that may block (for example, the invocation of a Servlet to handle an HTTP request).
+
+A task is typically a `Runnable` object that may implement `org.eclipse.jetty.util.thread.Invocable` to indicate the behavior of the task (in particular, whether the task may block or not).
+
+Once a task has been produced, it may be consumed using these strategies:
+
+* `Produce-Consume`
+* `Produce-Execute-Consume`
+* `Execute-Produce-Consume`
+
+[[pg-arch-threads-execution-strategy-pc]]
+===== Produce-Consume
+In the `Produce-Consume` mode, the producer thread loops to produce a task that is run directly by the `Producer Thread`.
+
+[plantuml]
+----
+skinparam backgroundColor transparent
+
+compact concise "Producer Thread" as PT
+hide time-axis
+
+@PT
+0 is T1 #lightgreen
+1 is "Run T1" #dodgerblue
+5 is T2 #lightgreen
+6 is "Run T2" #dodgerblue
+8 is T3 #lightgreen
+9 is "Run T3" #dodgerblue
+12 is {hidden}
+----
+
+If the task is a NIO selection event, then this model is the thread-per-selector model which is very CPU core cache efficient, but suffers from the link:http://en.wikipedia.org/wiki/Head-of-line_blocking[head-of-line blocking]: if one of the tasks blocks or runs slowly, then subsequent tasks cannot be produced (and therefore cannot be consumed either) and will pay in latency the cost of running previous, possibly unrelated, tasks.
+
+[[pg-arch-threads-execution-strategy-pec]]
+===== Produce-Execute-Consume
+In the `Produce-Execute-Consume` mode, the `Producer Thread` loops to produce tasks that are submitted to a `java.util.concurrent.Executor` to be run by ``Worker Thread``s different from the `Producer Thread`.
+
+[plantuml]
+----
+skinparam backgroundColor transparent
+
+compact concise "Producer Thread" as PT
+compact concise "Worker Thread 1" as WT1
+compact concise "Worker Thread 2" as WT2
+compact concise "Worker Thread 3" as WT3
+hide time-axis
+
+@PT
+0 is T1 #lightgreen
+1 is T2 #lightgreen
+2 is T3 #lightgreen
+3 is T4 #lightgreen
+4 is {hidden}
+
+@WT1
+1 is "Run T1" #dodgerblue
+5 is {hidden}
+
+@WT2
+2 is "Run T2" #dodgerblue
+4 is "Run T4" #dodgerblue
+8 is {hidden}
+
+@WT3
+3 is "Run T3" #dodgerblue
+6 is {hidden}
+----
+
+The `Executor` implementation typically adds the task to a queue, and dequeues the task when there is a worker thread available to run it.
```

Comment:
```suggestion
The `Executor` implementation typically adds the task to a queue, and dequeues the task when there is a worker thread available to run it. Configurations that support virtual threading are available that avoid such task queues.
```

Reply:
I don't think that is true. A virtual thread is "queued" as a continuation into the scheduler, typically the `ForkJoinPool`, so the generic sentence works for virtual thread too.

Reply:
But... But... Loom is magic! It is not bound by such concerns!

But seriously, i think a mention here that virtual threads are an alternative to thread pooling would be good

Reply:
Loom is trying to make it so that you don't care about those nuances and details.

That's what it seems like (at least) when you read their own documentation and blog posts about it.

Reply:
This paragraph is not the right section to introduce virtual threads.
I'm just explaining PEC, and how it works.
It is when PEC is used in the context of AdaptiveExecutionStrategy that virtual threads can be used, so it is explained there.

Reply:
Then leave out the typically explanation of what an executor does.